### PR TITLE
Correction to cost calculation

### PR DESCRIPTION
**Pull request by regan-huff**
Addresses https://github.com/allenai/nora-issues-research/issues/138

This change makes two adjustments to how we call litellm.cost_per_token.

First, it includes input_tokens_cache_write as cache_creation_input_tokens.

Second, it includes both cache_read and cache_write tokens as part of the total prompt_tokens.

The LiteLLM code considers total_tokens to be prompt_tokens + continuation_tokens:
from litellm.cost_calculator.cost_per_token:
```
usage_block = Usage(
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
            total_tokens=prompt_tokens + completion_tokens,
            cache_creation_input_tokens=cache_creation_input_tokens,
            cache_read_input_tokens=cache_read_input_tokens,
        )
```

For that to match our total_token count, prompt_tokens would need to include both input_tokens_cache_write and input_tokens_cache_read from our usage object:

```
"usage": {
              "input_tokens": 7,
              "output_tokens": 123,
              "total_tokens": 21877,
              "input_tokens_cache_write": 1101,
              "input_tokens_cache_read": 20646,
              "reasoning_tokens": null
            }
```

With these changes, the model costs calculated for the example in the original ticket come out positive:
```
  "model_costs": [
        0.3758097,
        0.2345571,
        0.1014348,
        0.15810795,
        0.17482694999999998,
        0.24655454999999998,
        0.14110215,
        0.20787885,
        0.14684115000000003,
        0.2512665
      ]
```

### CODE REVIEW

File: src/agenteval/log.py [hunk 1]
```diff
@@ -41,12 +41,15 @@ def compute_model_cost(model_usages: list[ModelUsageWithName]) -> float:
     """
     total_cost = 0.0
     for model_usage in model_usages:
+        cache_read_tokens = model_usage.usage.input_tokens_cache_read
+        cache_write_tokens = model_usage.usage.input_tokens_cache_write
         try:
             prompt_cost, completion_cost = cost_per_token(
                 model=model_usage.model,
-                prompt_tokens=model_usage.usage.input_tokens,
+                prompt_tokens=model_usage.usage.input_tokens + cache_read_tokens + cache_write_tokens,
```
Comment (mdarcy220) [inline src/agenteval/log.py:49]: Can you elaborate on how we know this is the right fix and maybe leave a code comment so we don't forget or accidentally revert it later?  I think it would also be good to test with a few different models to make sure we don't have corner cases like e.g. if openai cache tokens are reported differently than anthropic ones.  I suspect there's a reason litellm is interpreting the numbers the way it does and want to make sure we don't miss it.

Comment (mdarcy220) [reply src/agenteval/log.py:49]: > the advantage here would be that litellm could treat these cache costs differently

I don't understand; the cache token counts are still being passed below, so shouldn't litellm handle them the same way whether they're passed as Usage or as parameters?

Comment (jbragg) [reply src/agenteval/log.py:49]: In the solution I posted, `cost_per_token` would take the full usage object and interpret it. In the current solution in this PR, I see `prompt_tokens=model_usage.usage.input_tokens + cache_read_tokens + cache_write_tokens` which treats cache tokens like input tokens

Comment (jbragg) [reply src/agenteval/log.py:49]: Ah I didn't see the cache tokens being passed below. I don't know what litellm does, but regardless
- It seemed strange to me for the user to need to sum cache tokens in as prompt tokens
- I wanted to point out the alternative possibility of passing in a usage object

Comment (regan-huff) [reply src/agenteval/log.py:49]: @jbragg you can see this mention in the litellm cost_per_token function, and there's other evidence that they are calculating cost for "cache hit" tokens differently for anthropic models
```
### PROMPT CACHING PRICING ### - used for anthropic
    cache_creation_input_tokens: Optional[int] = 0,
    cache_read_input_tokens: Optional[int] = 0,
 ```
 
 I saw that you had another ticket open with the question about reasoning tokens. I do see litellm using them here:
 https://github.com/BerriAI/litellm/blob/main/litellm/litellm_core_utils/llm_cost_calc/utils.py#L336-L343
 
 I will try to compare results from passing the usage information in different ways.

**Comment by regan-huff** [comment]
Now understanding better that openai and anthropic (currently the only providers with prompt cache pricing) treat input_tokens_cache_write differently, with openai ignoring it and anthropic charging extra for cache writes. 

The usage results from these two providers are also calculated differently, with the openai input_tokens value including input_tokens_cache_read and the anthropic input_tokens value excluding both cache read and cache write tokens:

```
{
            "model": "gpt-4o-mini-2024-07-18",
            "usage": {
              "input_tokens": 15411,
              "output_tokens": 154,
              "total_tokens": 15565,
              "input_tokens_cache_write": null,
              "input_tokens_cache_read": 11648,
              "reasoning_tokens": 0
            }
```

```

 {
            "model": "claude-3-5-haiku-20241022",
            "usage": {
              "input_tokens": 6,
              "output_tokens": 112,
              "total_tokens": 11085,
              "input_tokens_cache_write": 1191,
              "input_tokens_cache_read": 9776,
              "reasoning_tokens": null
}
```

File: src/agenteval/log.py [hunk 2]
```diff
@@ -41,12 +41,19 @@ def compute_model_cost(model_usages: list[ModelUsageWithName]) -> float:
     """
     total_cost = 0.0
     for model_usage in model_usages:
+
+        prompt_tokens = model_usage.usage.input_tokens
+        # anthropic results split total tokens across input, output, cache write, and cache read values
+        if model_usage.usage.total_tokens != prompt_tokens + model_usage.usage.output_tokens:
+            prompt_tokens += (model_usage.usage.input_tokens_cache_write + model_usage.usage.input_tokens_cache_read)
```
Comment (mdarcy220) [inline src/agenteval/log.py:48]: Could probably re-check against total_tokens after this increment to be extra safe

**Comment by regan-huff** [comment]
With the current code in this PR, a hand calculation of cost for one litqa sample with openai matches the returned litellm cost exactly.

Hand calculation of anthropic cost for one litqa sample is significantly lower than the returned litellm cost.

This goes a bit deep in the weeds in the litellm code, but for interested parties, what I see happening is that when the litellm code reconstructs the Usage block [here](https://github.com/BerriAI/litellm/blob/main/litellm/cost_calculator.py#L168-L173) it makes some assumptions that don't quite hold for the Anthropic case.

```
usage_block = Usage(
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
            total_tokens=prompt_tokens + completion_tokens,
            cache_creation_input_tokens=cache_creation_input_tokens,
            cache_read_input_tokens=cache_read_input_tokens,
        )
```

If I set prompt_tokens = input_tokens + cache_read_input_tokens, I get a result that matches the hand calculated result (but in that case, total_tokens does not equal prompt_tokens + completion_tokens).

I think Jonathan's potential approach of passing in a litellm.Usage object will be needed to handle reasoning tokens, which it sounds like we want to do, so I'm inclined to make this correction by passing in a litellm.Usage object from agent-eval rather than trying to hack what I pass in to get the result I want from litellm's reconstruction of the Usage object.

**Comment by mdarcy220** [comment]
> If I set prompt_tokens = input_tokens + cache_read_input_tokens, I get a result that matches the hand calculated result (but in that case, total_tokens does not equal prompt_tokens + completion_tokens).

Should we report this as a bug to `litellm`?  It sounds like `total_tokens` should be `prompt_tokens + completion_tokens + cache_creation_input_tokens` based on the way it is currently set up

**Comment by mdarcy220** [comment]
> Note that I believe @mdarcy220 will be calling these from a provider different from `deepseek` like azure/togetherai

Which I think is yet another reason we should consider trying to standardize everything to litellm via a custom ModelApi; not sure exactly how much work it would be but if it would take care of name standardization and cost measurement gotchas, it could be worth it.

File: src/agenteval/log.py [hunk 3]
```diff
@@ -37,24 +38,52 @@ def collect_model_usage(sample: EvalSample) -> list[ModelUsageWithName]:
...
         try:
+            # input tokens count includes any cached tokens
+            if input_tokens == model_usage.usage.total_tokens - model_usage.usage.output_tokens:
+                text_tokens = input_tokens - cache_read_input_tokens
+                prompt_tokens = input_tokens
+
+            # (anthropic) input tokens count excludes cache read and cache write tokens
+            elif input_tokens == model_usage.usage.total_tokens - output_tokens - cache_read_input_tokens - cache_write_input_tokens:
+                text_tokens = input_tokens
+                prompt_tokens = input_tokens + cache_read_input_tokens + cache_write_input_tokens
```
Comment (mdarcy220) [inline src/agenteval/log.py:61]: This is actually a bit different than what I understood from your earlier [comment](https://github.com/allenai/agent-eval/pull/11#issuecomment-2957368934):

> Hand calculation of anthropic cost for one litqa sample is significantly lower than the returned litellm cost.
> If I set prompt_tokens = input_tokens + cache_read_input_tokens, I get a result that matches the hand calculated result 

So should this be `prompt_tokens = input_tokens + cache_read_input_tokens`?  As long as the result matches hand calculation it's all good, just making sure I'm not getting my wires crossed.

**Review by mdarcy220 — APPROVED** [review]
> @mdarcy220 Apologies for the circuitous journey.

On the contrary, thanks for looking so deep into this.  I'm really glad we're learning about this stuff now and not after we release...

> > Should we report this as a bug to litellm?
> 
> I see it as more a problem with the translation between the inspect_ai ModelUsage and the litellm Usage. To me the root inconsistency is that inspect_ai.ModelUsage.input_tokens means something different for different models. I like that this current solution makes that translation as visible and explicit as possible, but see what you think.

I agree that the semantics of `prompt_tokens == uncached_input_tokens + cache_read_input_tokens` is more a design decision than a bug, but I feel like if they are calculating `total_tokens` as `prompt_tokens + completion_tokens` when in fact it should be `prompt_tokens + cache_creation_input_tokens + completion_tokens`, that seems like a bug to me (surely `total_tokens` is intended to include ALL the input and output tokens; the main value of this number imo is determining whether you need to do history compression for subsequent steps).

But it's possible my understanding of the above is wrong; at least, based on the below code, `prompt_tokens` actually does equal `uncached_input_tokens + cache_read_input_tokens + cache_creation_input_tokens` (see comment) so perhaps I misunderstood the discussion.

But I think the updated solution looks good overall (just worth cross-checking against the earlier discussion on the `prompt_tokens` calculation for anthropic); thanks!

**Comment by regan-huff** [comment]
@mdarcy220 Thank you for your review! I left out a bit of information that might have been helpful.

Within the litellm cost calculation, they use the base input token price for the "text_token" count and the cache read price for the "cached_token" count. They then add a cost for the input_token_cache_creation tokens, which cost extra in the Anthropic case.

There are two paths for calling litellm.cost_per_token. Path 1 is if you call it as we were doing:
```
prompt_cost, completion_cost = cost_per_token(
                model=model_usage.model,
                prompt_tokens=model_usage.usage.input_tokens,
                completion_tokens=model_usage.usage.output_tokens,
                cache_read_input_tokens=model_usage.usage.input_tokens_cache_read,
            )
```
In this case, the litellm code reconstructs a litellm.Usage object. For what we were passing in, it was setting text_tokens [here](https://github.com/BerriAI/litellm/blob/main/litellm/litellm_core_utils/llm_cost_calc/utils.py#L242-L243):
```
## EDGE CASE - text tokens not set inside PromptTokensDetails
    if text_tokens == 0:
        text_tokens = usage.prompt_tokens - cache_hit_tokens - audio_tokens
```
Since we were setting prompt_tokens = input_tokens, in the Anthropic case this was often coming out negative when the cache was used (and we also weren't adding in the cost for cache write tokens). One way to hack this in the Anthropic case is to set prompt_tokens = input_tokens + input_tokens_cache_read, but that doesn't reflect reality very well, it just allows this existing codepath to work.

Path 2 for calling cost_per_token is to pass in a litellm.Usage object. This allows us to set text_tokens and cached_tokens explicitly. We can then set prompt_tokens = input_tokens + input_tokens_cache_read + input_tokens_cache_write without getting a negative value for text_tokens (because the litellm code doesn't have to reconstruct a Usage object, it already has one).

**Comment by regan-huff** [comment]
Exactly!
